### ISSUE TITLE
MethodOrderer.Random should print seed if none is configured to allow reproducibility

### ISSUE THREAD

**Issue by aepfli** [junit-team/junit5#1805]
MethodOrderer.Random is using `System.nanoTime()` as a seed to generate a random order. There is also a configuration property to allow reproducability by providing seed. It might be hard to determine the value of `System.nanoTime()` after execution, when there is no seed configured. Hence that it would be beneficiary to print the seed to the log with a message like `Using 'System.nanoTime()' as seed for ordering with value [%s].`.

Additionally the seed is right now computed for each test class, which makes it also hard to reproduce multiple tests at once. So the seed generation should be handled during the static initialization of `MethodOrderer.Random` so each test is utilizing the same seed.

Based on the current implementation i would suggest to use "logger.config" as it is also used if a seed is configured via property. like:
```
logger.config(() -> String.format("Using 'System.nanoTime()' as seed for ordering with value [%s].", seed)); 
```

Sidenote: Initially suggested by @nicolaiparlog during his live exploration of junit5 parallel execution.

## Deliverables

- [x] Log if random seed is generated via System.nanoTime()
- [x] provide a configuration parameter to swap default execution mode for `MethodOrderer.Random`
- [x] Appropriate test cases for changes

**Comment by aepfli** [comment]
i would also provide an implementation if this is accepted

**Comment by marcphilipp** [comment]
I think that's a good idea.

@sbrannen WDYT?

**Comment by sbrannen** [comment]
What do I think?

Well... I think it's a great idea... especially since I had the idea a long time and actually intended to implement it before 5.4 GA.

So... oops.... sorry about forgetting to do that. 😉

**Comment by sbrannen** [comment]
> i would also provide an implementation if this is accepted

@aepfli, feel free to submit a PR.

**Comment by sbrannen** [comment]
While we're on the topic, let me expound on my original idea.

The reason I didn't just log the seed in every case is that I was pondering whether it would make sense to compute the random seed _globally_ and reuse the same one across all randomly ordered test classes within the current test plan.

Rationale: currently a new random seed is generated for every test class configured with `MethodOrderer.Random` (within `orderMethods()`), and that results in a different random seed for each test class. So if we were to log that for every single test class, the user would have to track down (in the logs) the exact seed used for a given test class. That's fine if one only wants to rerun that single test class with the same seed, but if one needs/wants to rerun several test classes it's simply not possible to supply the random seed to use per test class (in the same execution of the test plan).

So one idea I had was to precompute the "default" random seem -- for example, during static initialization of the `MethodOrderer.Random` class -- and log its value only once.

Thoughts?

**Comment by aepfli** [comment]
+1 - sounds like a legit and good plan.

**Comment by marcphilipp** [comment]
+1

**Comment by aepfli** [comment]
@sbrannen updated the description -> i do have another question:

In the code the MethodOrderer will be using `SAME_THREAD`, if a seed is provided via configuration. Might this cause any troubles? I am not sure how `java.util.Random` will behave in a multi thread environment, and if we really achieve predictability in execution Order with a static seed in multiple threads. because we still might execute the tests in a different order based on distribution over threads.
Should we maybe also provide a configuration property like `junit.jupiter.execution.order.random.concurrent` and document that this might cause problematic behaviour with concurrent setup?

I would highly appreciate some input regarding this.

**Comment by sormuras** [comment]
Use [`ThreadLocalRandom`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadLocalRandom.html) instead?

**Comment by aepfli** [comment]
Maybe Random is not the issue.

My problem relates to this override:
https://github.com/junit-team/junit5/blob/5e41ebe612fd67e905f96f1dd3184a071b65be17/junit-jupiter-api/src/main/java/org/junit/jupiter/api/MethodOrderer.java#L238-L254

This sounds too me reasonable, to set the execution to same thread. But this also means, to make the execution reproducable, so but this would require to run the tests also in the same thread, if i just use a static seed,. A static seed might provide the same order, but it will not ensure, that the tests will be distributed into different threads the same way on every run. So should i care about this now, or is a note in the docs enough for this. It means that reproducability might be achieved, but it is not guaranteed, when you want to rerun a test and try to provide the same seed again.

**Comment by sbrannen** [comment]
Good questions, @aepfli!

I guess we essentially have two options.

1. Keep the concurrency semantics for a default seed _as is_, but provide a configuration parameter (as you proposed) to ensure execution in the same thread for repeatability.
2. Change the concurrency semantics for a default seed so that we always use `ExecutionMode.SAME_THREAD` for randomly ordered tests.

For the latter, we could also opt to but provide a configuration parameter to ensure execution in the same thread for repeatability.

I'm leaning toward # 1.

Rationale:

- If developers pick random, they probably actually want it random _and_ concurrent (if desired). If they are already configuring concurrency, that means that they are (hopefully) already putting thought into that, and they could simultaneously reason about whether they'd like to configure concurrency for random tests as well.
- If developers supply an explicit seed, they already know they're opting out of concurrent execution.

@junit-team/junit-lambda, thoughts?

**Comment by sbrannen** [comment]
> Use [`ThreadLocalRandom`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadLocalRandom.html) instead?

I don't think that would have any effect.

We use the instance of `Random` during the discovery phase, which is single threaded -- right?

**Comment by aepfli** [comment]
i am also tending toward # 1.

hence that and the fact that i was a little bit bored yesterday, i started the implementation. there are just a few things i do have problems.

1. i cant test the static initializer blog, and if the logging is working properly
2. i did not find a way to verify, which execution mode was used. There is the test class `OrderedMethodTests.java` which i extended/adapted. But i see no way of access the execution mode, to back this implementation with a unittest.

I know, this is right now still in discussion. but as soon as we agreed on an approach, can i open a [WIP]-pullrequest to get some feedback regarding that?

**Comment by sbrannen** [comment]
>     1. i cant test the static initializer blog, and if the logging is working properly

That should be achievable by using a custom `ClassLoader` and `@TrackLogRecords`.

>     2. i did not find a way to verify, which execution mode was used. There is the test class `OrderedMethodTests.java` which i extended/adapted. But i see no way of access the execution mode, to back this implementation with a unittest.

You can track the name of the current thread in each test -- for example, in a `static` `List<String` -- and assert the results in an _integration test_.

> I know, this is right now still in discussion. but as soon as we agreed on an approach, can i open a [WIP]-pullrequest to get some feedback regarding that?

Yeah, sure.

**Comment by marcphilipp** [comment]
**Team Decision:** Always use `ExecutionMode.SAME_THREAD` for randomly ordered tests.

**Comment by aepfli** [comment]
Nice: I will adapt the code and open the pull request this weekend

**Comment by wkorando** [comment]
Should the logging level be set to a higher priority (INFO maybe?)? CONFIG doesn't print by default with the Java util logger, which means the seed value that might had exposed a problem in the test suite and/or underlying application will likely not had been printed to console.

**Comment by sbrannen** [comment]
@wkorando, feel free to raise a new issue regarding the log level, and we'll do our best to discuss it before release 5.5 GA.
